File: db4oj/core/src/com/db4o/inside/btree/algebra/BTreeAlgebra.java
/* Copyright (C) 2006 db4objects Inc. http://www.db4o.com */

package com.db4o.inside.btree.algebra;

import com.db4o.foundation.*;
import com.db4o.inside.btree.*;


/**
 * @exclude
 */
class BTreeAlgebra {

	public static BTreeRange intersect(BTreeRangeUnion union, BTreeRangeSingle single) {
		throw new com.db4o.foundation.NotImplementedException();
	}

	public static BTreeRange intersect(BTreeRangeUnion union1, BTreeRangeUnion union2) {
		throw new com.db4o.foundation.NotImplementedException();
	}

	public static BTreeRange intersect(BTreeRangeSingle single1, BTreeRangeSingle single2) {
		BTreePointer first = BTreePointer.max(single1.first(), single2.first());
		BTreePointer end = BTreePointer.min(single1.end(), single2.end());
		return single1.newBTreeRangeSingle(first, end);
	}

	public static BTreeRange union(final BTreeRangeUnion union1, final BTreeRangeUnion union2) {
		final Iterator4 ranges = union1.ranges();
		BTreeRange merged = union2;
		while (ranges.moveNext()) {
			merged = merged.union((BTreeRange) ranges.current());
		}
		return merged;
	}

	public static BTreeRange union(final BTreeRangeUnion union, final BTreeRangeSingle single) {
		if (single.isEmpty()) {
			return union;
		}
		
		SortedCollection4 sorted = new SortedCollection4(BTreeRangeSingle.COMPARISON);
		sorted.add(single);		
		
		BTreeRangeSingle range = single;
		Iterator4 ranges = union.ranges();
		while (ranges.moveNext()) {
			BTreeRangeSingle current = (BTreeRangeSingle) ranges.current();
			if (canBeMerged(current, range)) {
				sorted.remove(range);
				range = merge(current, range);
				sorted.add(range);
			} else {
				sorted.add(current);
			}
		}
		
		if (1 == sorted.size()) {
			return (BTreeRange)sorted.singleElement();
		}
		
		return new BTreeRangeUnion(sorted);
	}
	
	public static BTreeRange union(final BTreeRangeSingle single1, final BTreeRangeSingle single2) {
		if (single1.isEmpty()) {
			return single2;
		}
		if (single2.isEmpty()) {
			return single1;
		}
		if (canBeMerged(single1, single2)) {
			return merge(single1, single2);
		}
		return new BTreeRangeUnion(new BTreeRangeSingle[] { single1, single2 });
	}
	
	private static BTreeRangeSingle merge(BTreeRangeSingle range1, BTreeRangeSingle range2) {
		return range1.newBTreeRangeSingle(
					BTreePointer.min(range1.first(), range2.first()),
					BTreePointer.max(range1.end(), range2.end()));
	}

	private static boolean canBeMerged(BTreeRangeSingle range1, BTreeRangeSingle range2) {
		return range1.overlaps(range2)
				|| range1.adjacent(range2);
	}
}

File: db4on/core/src/core/com/db4o/inside/btree/algebra/BTreeAlgebra.cs
namespace com.db4o.inside.btree.algebra
{
	/// <exclude></exclude>
	internal class BTreeAlgebra
	{
		public static com.db4o.inside.btree.BTreeRange Intersect(com.db4o.inside.btree.BTreeRangeUnion
			 union, com.db4o.inside.btree.BTreeRangeSingle single)
		{
			com.db4o.foundation.SortedCollection4 collection = NewBTreeRangeSingleCollection(
				);
			CollectIntersections(collection, union, single);
			return ToRange(collection);
		}

		public static com.db4o.inside.btree.BTreeRange Intersect(com.db4o.inside.btree.BTreeRangeUnion
			 union1, com.db4o.inside.btree.BTreeRangeUnion union2)
		{
			com.db4o.foundation.SortedCollection4 collection = NewBTreeRangeSingleCollection(
				);
			System.Collections.IEnumerator ranges = union1.Ranges();
			while (ranges.MoveNext())
			{
				com.db4o.inside.btree.BTreeRangeSingle current = (com.db4o.inside.btree.BTreeRangeSingle
					)ranges.Current;
				CollectIntersections(collection, union2, current);
			}
			return ToRange(collection);
		}

		private static void CollectIntersections(com.db4o.foundation.SortedCollection4 collection
			, com.db4o.inside.btree.BTreeRangeUnion union, com.db4o.inside.btree.BTreeRangeSingle
			 single)
		{
			System.Collections.IEnumerator ranges = union.Ranges();
			while (ranges.MoveNext())
			{
				com.db4o.inside.btree.BTreeRangeSingle current = (com.db4o.inside.btree.BTreeRangeSingle
					)ranges.Current;
				if (single.Overlaps(current))
				{
					collection.Add(single.Intersect(current));
				}
			}
		}

		public static com.db4o.inside.btree.BTreeRange Intersect(com.db4o.inside.btree.BTreeRangeSingle
			 single1, com.db4o.inside.btree.BTreeRangeSingle single2)
		{
			com.db4o.inside.btree.BTreePointer first = com.db4o.inside.btree.BTreePointer.Max
				(single1.First(), single2.First());
			com.db4o.inside.btree.BTreePointer end = com.db4o.inside.btree.BTreePointer.Min(single1
				.End(), single2.End());
			return single1.NewBTreeRangeSingle(first, end);
		}

		public static com.db4o.inside.btree.BTreeRange Union(com.db4o.inside.btree.BTreeRangeUnion
			 union1, com.db4o.inside.btree.BTreeRangeUnion union2)
		{
			System.Collections.IEnumerator ranges = union1.Ranges();
			com.db4o.inside.btree.BTreeRange merged = union2;
			while (ranges.MoveNext())
			{
				merged = merged.Union((com.db4o.inside.btree.BTreeRange)ranges.Current);
			}
			return merged;
		}

		public static com.db4o.inside.btree.BTreeRange Union(com.db4o.inside.btree.BTreeRangeUnion
			 union, com.db4o.inside.btree.BTreeRangeSingle single)
		{
			if (single.IsEmpty())
			{
				return union;
			}
			com.db4o.foundation.SortedCollection4 sorted = NewBTreeRangeSingleCollection();
			sorted.Add(single);
			com.db4o.inside.btree.BTreeRangeSingle range = single;
			System.Collections.IEnumerator ranges = union.Ranges();
			while (ranges.MoveNext())
			{
				com.db4o.inside.btree.BTreeRangeSingle current = (com.db4o.inside.btree.BTreeRangeSingle
					)ranges.Current;
				if (CanBeMerged(current, range))
				{
					sorted.Remove(range);
					range = Merge(current, range);
					sorted.Add(range);
				}
				else
				{
					sorted.Add(current);
				}
			}
			return ToRange(sorted);
		}

		private static com.db4o.inside.btree.BTreeRange ToRange(com.db4o.foundation.SortedCollection4
			 sorted)
		{
			if (1 == sorted.Size())
			{
				return (com.db4o.inside.btree.BTreeRange)sorted.SingleElement();
			}
			return new com.db4o.inside.btree.BTreeRangeUnion(sorted);
		}

		private static com.db4o.foundation.SortedCollection4 NewBTreeRangeSingleCollection
			()
		{
			return new com.db4o.foundation.SortedCollection4(com.db4o.inside.btree.BTreeRangeSingle
				.COMPARISON);
		}

		public static com.db4o.inside.btree.BTreeRange Union(com.db4o.inside.btree.BTreeRangeSingle
			 single1, com.db4o.inside.btree.BTreeRangeSingle single2)
		{
			if (single1.IsEmpty())
			{
				return single2;
			}
			if (single2.IsEmpty())
			{
				return single1;
			}
			if (CanBeMerged(single1, single2))
			{
				return Merge(single1, single2);
			}
			return new com.db4o.inside.btree.BTreeRangeUnion(new com.db4o.inside.btree.BTreeRangeSingle
				[] { single1, single2 });
		}

		private static com.db4o.inside.btree.BTreeRangeSingle Merge(com.db4o.inside.btree.BTreeRangeSingle
			 range1, com.db4o.inside.btree.BTreeRangeSingle range2)
		{
			return range1.NewBTreeRangeSingle(com.db4o.inside.btree.BTreePointer.Min(range1.First
				(), range2.First()), com.db4o.inside.btree.BTreePointer.Max(range1.End(), range2
				.End()));
		}

		private static bool CanBeMerged(com.db4o.inside.btree.BTreeRangeSingle range1, com.db4o.inside.btree.BTreeRangeSingle
			 range2)
		{
			return range1.Overlaps(range2) || range1.Adjacent(range2);
		}
	}
}
